torchqdynamics/mesolve.py

```diff
@@ -69,7 +88,62 @@ def forward_adjoint(self, t, dt, phi):
 
 class MERouchon2(MERouchon):
     def forward(self, t, dt, rho):
-        raise NotImplementedError
+        """Compute rho(t+dt) using a Rouchon method of order 2.
+
+        NOTE: For fast time-varying Hamiltonians, this method is not order 2 because the
+        second-order time derivative term is neglected. This term could be added in the
+        zero-th order Kraus operator if needed, as M0 += -0.5j * dt**2 * \dot{H}.
+        """
+        # non-hermitian Hamiltonian at time t
+        H_nh = self.H(t) - 0.5j * self.sum_nojump
+
+        # build time-dependent Kraus operators
+        M0 = self.I - 1j * dt * H_nh - 0.5 * dt**2 * H_nh @ H_nh
+        M1s = 0.5 * sqrt(dt) * (self.jump_ops @ M0 + M0 @ self.jump_ops)
+
+        # compute rho(t+dt)
+        rho_ = kraus_map(rho, M1s)
+        rho = kraus_map(rho, M0) + rho_ + 0.5 * kraus_map(rho_, M1s)
+        return rho / btrace(rho)
 
     def forward_adjoint(self, t, dt, phi):
         raise NotImplementedError
+
+
+def kraus_map(rho, *operators):
+    """Compute the application of a Kraus map on an input density matrix.
+
+    This is equivalent to `torch.sum(op @ rho @ op.adjoint())`.
+
+    Kraus operators are batched together such that only a single matrix operation is
+    called. The use of einsum yields better performances on large matrices, but may
+    cause a small overhead on smaller matrices (N <~ 50).
+    """
+    if any(op.ndim > 3 or op.ndim < 2 for op in operators):
+        raise ValueError(
+            'Kraus operators should be 2-D tensors, or 3-D tensors if batched.'
+        )
+
+    ops_batch = torch.cat(
+        tuple(op if op.ndim == 3 else op.unsqueeze(0) for op in operators)
+    )
+    return torch.einsum('kin,...nm,kjm->...ij', ops_batch, rho, ops_batch.conj())
```

Comment:
Niiiice! I love the benchmark, einsum is actually x2 faster on my CPU on this example, even with small matrix sizes. 🤓  I take the chance to give some insights of what I think is good practice for our codebase.

- First, we should be careful with this kind of tricks where the argument tensors can have different size, it feels easier at first but it's most of the time better to make strict functions that don't do fancy stuff in the background. We will use more pytorch tricks here and there (like `x[None,...]`), but the code will be very readable for those who know the basics.
- Then I think we should always indicate when we make a generic function working with tensors the dimensions of the arguments and the returned tensor (including batching). If we do so, I don't think we need to check the tensors dimensions everytime, we should trust the user.
- For the einstein summation, I like the `i, j, k, l,...` indices for the matrix multiplications and `n, m, ...` indices for the repeated dimensions to sum over. For example here writing `'mij,...jk,mkl ->...il'` we see nicely the `ij, jk, kl` matrix multiplication contraction and the `m` is just summed over.
- I think we want `adjoint` and not `conj`?

So I suggest:
```python
def kraus_map(rho, ops):
    """ ... """
    # rho: `(..., d, d)`
    # ops: `(n, d, d)`
    # -> (..., d, d)
    return torch.einsum('mij,...jk,mkl ->...il', ops, rho, ops.adjoint())
```
and in e.g. Rouchon1
```python
ops = torch.cat((M0[None,...], sqrt(dt) * self.jump_ops))
rho = kraus_map(rho, ops)
```